### ISSUE TITLE
A few problems with FixedAdjacencyGraphConvolution

### ISSUE THREAD

**Issue by davidxujiayang** [stellargraph/stellargraph#1439]
Hi, 
Great to see that spatio-temporal networks are also being implemented in this library. I have a few problem with the gcn_lstm layer as follows
1）Why is the input dim limited to (batch size, sequence length, number of nodes) in FixedAdjacencyGraphConvolution? I think it should be 4D: (batch size, sequence length, number of nodes, feature vector size)
2)  The why converting sparse adjacency matrix into a nontrainable dense matrix makes the memory consumption huge. I assume that this is due to K.dot's not supporting batchsize>1. Have you considered using timedistributed wrapper? - limiting batch size to 1 to use sparse adj as in normal GCN layer or until it can be fixed in the same way there. But use timedistributed on the sequence length dimension. 
3) It seems that the FixedAdjacencyGraphConvolution in GraphConvolutionLSTM have to use a fixed number of channels (this is related to 1, here the number of channel seems to be set to t_steps, which is weird. The hidden state should have dimensions (batch size, sequence length, number of nodes, number of channels).

My understanding about the dimensions is based on the ConvLSTM, which is based on non-graph convolutions so things can not be the same. Looking forward to your reply

-David

**Comment by habiba-h** [comment]
Hi @davidxujiayang , thanks for checking out `GraphConvolutionLSTM` model. Thanks for the thorough feedback. We are working on improving the model. 
A quick response to the issues you mentioned. 

> 1）Why is the input dim limited to (batch size, sequence length, number of nodes) in FixedAdjacencyGraphConvolution? I think it should be 4D: (batch size, sequence length, number of nodes, feature vector size)

This method works for univariate timeseries and the sequence dimension captures the feature dimension. The 4th dimension is not needed here. Also, LSTM works with 3 dimensions. We thought about encoding the network structure as features but couldn't see how do it logically. If you have any ideas around that please let us know. 
 


> 1. The why converting sparse adjacency matrix into a nontrainable dense matrix makes the memory consumption huge. I assume that this is due to K.dot's not supporting batchsize>1.

You answered it yourself, for the end-to-end training, the Graph convolutions have to be performed for entire training batch. Making the batch size equal to one, is not practical. 




> 1. Actually it seems that the adjacency matrix and features are also sent in separatedly into the normal GraphConvolution layer, if a wrapper works maybe there is no need to develop a seperate one for fixedadjacency unless for performance reasons? Just replicating the sparse adjacency matrix into a temporal dimension should work. This may sounds silly but it should still be much more memory friendly than converting to dense, unless the number of time steps is on the same order as number of nodes. If more work can be put into modifying the timedistributed wrapper s.t. it takes only one adj matrix and multiplies it to every time step of the feature matrix, then the repetition can be avoided.

My experience with `keras`  `TimeDistributed` wrapper layer is with `ConvLSTM`,  using graph convolution instead of simple convolution is not straightforward. So I thought to start with a simple solution. I am looking into using the timedistributed wrapper and passing the convolved adjacency matrix direct into the LSTM block. 



> 2\. It seems that the FixedAdjacencyGraphConvolution in GraphConvolutionLSTM have to use a fixed number of channels (this is related to 1, here the number of channel seems to be set to t_steps, which is weird. The hidden state should have dimensions (batch size, sequence length, number of nodes, number of channels).

As I mentioned earlier, for the univariate timeseries, there is one feature which is captured in the sequence dimension. So there is no need for a 4th dimension. For graph convolution, we need the graph (adjacency) and the features (timeseries in this case). These graph convolution are done for each batch>1 of inputs, the output of which gets distributed over the LSTM layer. 

I might be missing something here. Please, elaborate on it, if the explanation isn't sufficient. 

I will be looking into the timedistributed wrapper and if there is a way to avoid creating dense matrix with nontrivial batch size. 

Looking forward to your further thoughts on this method.

Thanks.
Habiba

**Comment by timpitman** [comment]
Title updated as #1528 is now a separate task

**Comment by davidxujiayang** [comment]
Hi @habiba-h, thanks for your reply. I modified the layer and model and it works in my case. Attached are the files, some dependencies may look messy, sorry for that. It should be noted that in my model the LSTM works in a local way, that is, the future state for each local node is only based on its own history. This is achieved by shuffle and use timedistributed on the spatial dimension and shuffle back, which is irrelevant to this channel although the wrapper is used. 

I also added a hard-coded work around for the limited batch size of 1. The model/layer should work with arbitrary fixed batch size, which has to be passed as a parameter fixed_batch_size when the model is built. 

Best,
David

[fixedAdjacency.zip](https://github.com/stellargraph/stellargraph/files/4590073/fixedAdjacency.zip)

**Comment by habiba-h** [comment]
Hi @davidxujiayang, thanks for sharing your attempt at the modification. This would be really helpful! I will have a look and update you once I have  made the changes and tested them. Thanks for the contribution :-)!

**Comment by habiba-h** [comment]
Hi @davidxujiayang , thanks once again for your suggestions and input for modifying the `GraphConvolutionLSTM`. I tested the `TimeDistributed` wrapper over the `FixedAdjacencyGraphConvolution` and the `LSTM` layer. However, I could only make it work with the batch size 1, which made the training extremely slow and worsened the performance. 

If you would like to contribute directly to the library by incorporating the modifications you suggested, please, feel free to do so.

**Comment by davidxujiayang** [comment]
Hi @habiba-h, would you like to give tensorflow map_fn a try? It can be used like `h_graph = map_fn(lambda x: K.dot(A, x) , features)`. This can deal with arbitrary batch size, which does not need to be predefined when building the model, which is needed if a loop is used rather than map_fn.

However, although this  deals with flexible batch size, the dot computation is still done in a looping manner, thus it will not see any acceleration in the graph convolution part.

Best,
David

**Comment by habiba-h** [comment]
Hi @davidxujiayang , thanks for the suggestion. I will look into `map_fn` and see how it performs. Will share my thoughts on this once I have tried it.
Thanks for staying involved and great suggestions!
Best, 
Habiba
